### PR TITLE
coreutils: add tac command

### PR DESCRIPTION
**Pull request by MaheshBharadwaj**
#162: Coreutils:  add **tac** command

The tac command displays the contents of the file in reverse order after splitting contents based on a separator(by defualt \n)

## Output
Screenshot from 2020-10-18 13-32-51

## Output on ubuntu
Screenshot from 2020-10-18 13-35-18

## Known issues:
- Separator is assumed to be a single character
- Doesn't work on STDIN

### CODE REVIEW

File: applications/coreutils/tac.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        if (handle_has_error(stream))
+        {
+            return handle_get_error(stream);
+        }
+        String tmp(buffer, read);
+        s += tmp;
+    }
+
+    String *tmp = new String();
```
Comment (sleepy-monax) [inline applications/coreutils/tac.cpp:61]: `String str{}` could work too :)

File: applications/coreutils/tac.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        String tmp2(s[i]);
+        if (s[i] == sep || i == s.length() - 1)
+        {
+            if (!before)
+            {
+                *tmp += tmp2;
+            }
+            lines.push_back((*tmp));
+            delete tmp;
+            tmp = new String();
```
Comment (sleepy-monax) [inline applications/coreutils/tac.cpp:73]: tmp = {}

File: applications/coreutils/tac.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    size_t read;
+    char buffer[1024];
+    char sep = (separator == nullptr ? '\n' : separator[0]);
+
+    String s;
+    Vector<String> lines(20);
+
+    while ((read = stream_read(stream, &buffer, 1024)) != 0)
+    {
+        if (handle_has_error(stream))
```
Comment (sleepy-monax) [inline applications/coreutils/tac.cpp:53]: This error handler will never be reached because when failing read() returns 0

File: applications/coreutils/tac.cpp (hunk 1 above)
Comment (caelwithcats) [inline applications/coreutils/tac.cpp:61]: The String type doesn't need to be created on the heap

**Review by caelwithcats — CHANGES_REQUESTED** [review]

File: applications/coreutils/tac.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    for (unsigned int i = lines.count(); i >= 1; --i)
+    {
+        stream_write(out_stream, lines[i - 1].cstring(), lines[i - 1].length());
+        if (handle_has_error(out_stream))
+        {
+            return ERR_WRITE_STDOUT;
+        }
+    }
+
+    stream_flush(out_stream);
```
Comment (sleepy-monax) [inline applications/coreutils/tac.cpp:94]: out_stream is already flushed automatically when the process exit

File: applications/coreutils/tac.cpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+{
+
+    __cleanup(stream_cleanup) Stream *stream = stream_open(path, OPEN_READ);
+
+    if (handle_has_error(stream))
+    {
+        return handle_get_error(stream);
+    }
+    FileState stat = {};
+    stream_stat(stream, &stat);
```
Comment (sleepy-monax) [inline applications/coreutils/tac.cpp:42]: The stat variable is unused

File: applications/coreutils/tac.cpp (hunk 1 above)
Comment (caelwithcats) [reply applications/coreutils/tac.cpp:61]: You could just do this:
```
tmp = "";
```

File: applications/coreutils/tac.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+Result tac(const char *path)
+{
+
+    __cleanup(stream_cleanup) Stream *stream = stream_open(path, OPEN_READ);
+
+    if (handle_has_error(stream))
+    {
+        return handle_get_error(stream);
+    }
```
Comment (sleepy-monax) [inline applications/coreutils/tac.cpp:40]: Why not open the file outside of the tac() function and pass the stream as a parameter. This will give you stdin support for free :)

**Review by sleepy-monax — CHANGES_REQUESTED** [review]
You are doing a lot of strings concatenation, please use libutils/StringBuilder instead. It's way faster than the + operator.

**Review by Keyboard-Slayer — CHANGES_REQUESTED** [review]
Thank you for taking time to the project :+1: 
Can you take note of those changes please ?

File: applications/coreutils/tac.cpp [hunk 7]
```diff
@@ -91,33 +76,47 @@ Result tac(const char *path)
...
         return PROCESS_SUCCESS;
     }
 
-    Result result;
-
     for (int i = 1; i < argc; i++)
     {
-        result = tac(argv[i]);
+        path = path_create(argv[i]);
+        __cleanup(stream_cleanup) Stream *stream = stream_open(path_as_string(path), OPEN_READ);
```
Comment (sleepy-monax) [inline applications/coreutils/tac.cpp:104]: stream_open(argv[i], OPEN_READ); is fine no need to creat a Path object.

File: applications/coreutils/tac.cpp [hunk 8]
```diff
@@ -29,60 +30,44 @@ static CommandLine cmdline = CMDLINE(
...
-    while ((read = stream_read(stream, &buffer, 1024)) != 0)
+    while ((read = stream_read(input_stream, buffer, 1024)) != 0)
     {
-        if (handle_has_error(stream))
-        {
-            return handle_get_error(stream);
-        }
-        String tmp(buffer, read);
-        s += tmp;
+        s.append(String(buffer, read));
```
Comment (sleepy-monax) [inline applications/coreutils/tac.cpp:44]: s.append(buffer, read); No need to create a string object :)

**Review by sleepy-monax — CHANGES_REQUESTED** [review]
Reading, the whole to memory and splitting it is a bit efficient. I think you should split the file as you are reading it.

File: applications/coreutils/tac.cpp [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+                         COMMANDLINE_NO_CALLBACK),
+    COMMANDLINE_OPT_END};
+
+static CommandLine cmdline = CMDLINE(
+    usages,
+    options,
+    "Concatenate and print lines of file(s) in reverse.",
+    "If no filename provided read from input stream");
+
+Result tac(const char *path)
```
[the remark this answers is not in the document]
Comment (sleepy-monax) [reply applications/coreutils/tac.cpp]: No need for the path object.

File: applications/coreutils/tac.cpp [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    if (handle_has_error(stream))
+    {
+        return handle_get_error(stream);
+    }
+    FileState stat = {};
+    stream_stat(stream, &stat);
+
+    size_t read;
+    char buffer[1024];
+    char sep = (separator == nullptr ? '\n' : separator[0]);
```
[the remark this answers is not in the document]
Comment (sleepy-monax) [reply applications/coreutils/tac.cpp:37]: why not do :
if (!separator) {
    separator = "\n";
}

File: applications/coreutils/tac.cpp [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    for (size_t i = 0; i < file_contents.length(); ++i)
+    {
+        if (file_contents[i] == sep || i == file_contents.length() - 1)
+        {
+            if (!before)
+            {
+                temp.append(file_contents[i]);
+            }
+
+            lines.push_back(temp.finalize());
```
Comment (Keyboard-Slayer) [inline applications/coreutils/tac.cpp:58]: There's maybe a more optimal way to do that :thinking: 
(between the beginning of the `for` loop to its end)

File: applications/coreutils/tac.cpp [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+                temp.append(file_contents[i]);
+            }
+        }
+        else
+        {
+            temp.append(file_contents[i]);
+        }
+    }
+
+    for (size_t i = lines.count(); i >= 1; --i)
```
Comment (Keyboard-Slayer) [inline applications/coreutils/tac.cpp:70]: You should use `i > 0`  instead of ` i >= 1`

File: applications/coreutils/tac.cpp [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+        result = tac(stream);
+
+        if (result != SUCCESS)
+        {
+            stream_format(err_stream, "%s: %s: %s", argv[0], argv[i], get_result_description(result));
+        }
+    }
+    path_destroy(path);
+
+    return PROCESS_SUCCESS;
```
Comment (Keyboard-Slayer) [inline applications/coreutils/tac.cpp:121]: It should return `PROCESS_FAILURE` when one of the function returns an error

File: applications/coreutils/tac.cpp (hunk 8 above)
Comment (MaheshBharadwaj) [reply applications/coreutils/tac.cpp:44]: i couldnt find a method which takes in char* and size_t

File: applications/coreutils/tac.cpp [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    usages,
+    options,
+    "Concatenate and print lines of file(s) in reverse.",
+    "If no filename provided read from input stream");
+
+Result tac(Stream *const input_stream)
+{
+    size_t read;
+    char buffer[1024];
+    char sep = (separator == nullptr ? '\n' : separator[0]);
```
Comment (Keyboard-Slayer) [inline applications/coreutils/tac.cpp:37]: You should do that insead
```c++
if (!separator) {
    separator = "\n";
}
```

and then use directly the string instead of the `char`

File: applications/coreutils/tac.cpp [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+    Vector<String> lines(20);
+
+    while ((read = stream_read(input_stream, buffer, 1024)) != 0)
+    {
+        s.append(String(buffer, read));
+    }
+
+    String file_contents = s.finalize();
+
+    for (size_t i = 0; i < file_contents.length(); ++i)
```
Comment (Keyboard-Slayer) [inline applications/coreutils/tac.cpp:49]: You should prefer `var++` and `var--` instead of `++var` and `--var`

**Review by Keyboard-Slayer — CHANGES_REQUESTED** [review]

File: applications/coreutils/tac.cpp (hunk 14 above)
Comment (MaheshBharadwaj) [reply applications/coreutils/tac.cpp:37]: Im checking for a character to split so idk how to check for substring, would strstr be ok?

File: applications/coreutils/tac.cpp [hunk 16]
```diff
@@ -29,60 +30,44 @@ static CommandLine cmdline = CMDLINE(
...
-    }
-    FileState stat = {};
-    stream_stat(stream, &stat);
-
     size_t read;
     char buffer[1024];
     char sep = (separator == nullptr ? '\n' : separator[0]);
 
-    String s;
+    StringBuilder s, temp;
```
[the remark this answers is not in the document]
Comment (MaheshBharadwaj) [reply applications/coreutils/tac.cpp:39]: separate as in place them in individual lines?

[the remark this answers is not in the document]
Comment (Keyboard-Slayer) [reply applications/coreutils/tac.cpp:39]: as individual lines

File: applications/coreutils/tac.cpp (hunk 14 above)
Comment (Keyboard-Slayer) [reply applications/coreutils/tac.cpp:37]: Tac seperate on string and not on char 

```
       -s, --separator=STRING
              use STRING as the separator instead of newline
```
(from the man page of `tac`)

for your question, you should ask to @sleepy-monax

Comment (sleepy-monax) [reply applications/coreutils/tac.cpp:37]: Yes we don't have anything better than strstr right now ;)

File: applications/coreutils/tac.cpp [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+    static char *start = NULL;
+
+    if(!start)
+    {
+        start = str;
+    }
+
+    if(!*start)
+    {
+        return NULL;
```
Comment (Keyboard-Slayer) [inline applications/coreutils/tac.cpp:51]: You should return `nullptr` instead of `null`

File: applications/coreutils/tac.cpp [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+        strcpy(separator, "\n");
+    }
+
+    while ((read = stream_read(input_stream, buffer, 1024)) != 0)
+    {
+        
+        buffer[read] = 0;
+        char *copy = strdup(buffer);
+        split = str_split(copy, separator);
+        while (split != NULL)
```
Comment (Keyboard-Slayer) [inline applications/coreutils/tac.cpp:96]: You use `nullptr` on line 79, you should use it here too

**Review by Keyboard-Slayer — CHANGES_REQUESTED** [review]
Woaaaw your code is very clean, there is still small issues with the usage of `NULL` instead of `nullptr`

File: applications/coreutils/tac.cpp (hunk 17 above)
Comment (MaheshBharadwaj) [reply applications/coreutils/tac.cpp:51]: also at the end of the function I presume, where im currently returning `NULL`

File: applications/coreutils/tac.cpp [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+    if (!separator)
+    {
+        separator = new char[2];
+        strcpy(separator, "\n");
+    }
+
+    while ((read = stream_read(input_stream, buffer, 1024)) != 0)
+    {
+
+        buffer[read] = 0;
```
Comment (sleepy-monax) [inline applications/coreutils/tac.cpp:93]: You have a buffer overflow here when read = 1024

File: applications/coreutils/tac.cpp [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+        char *copy = strdup(buffer);
```
Comment (sleepy-monax) [inline applications/coreutils/tac.cpp:94]: Why do you need to make a copy of the buffer? Also, you are not freeing this memory.

**Review by sleepy-monax — CHANGES_REQUESTED** [review]
This nearly ready to merge you have just 2 bugs to fix :)

**Comment by sleepy-monax** [comment]
@MaheshBharadwaj Do you feel it's ready to be merged? :)

**Comment by MaheshBharadwaj** [comment]
> @MaheshBharadwaj Do you feel it's ready to be merged? :)

I think it is ready :) is there any file i should test with to ensure it is working properly?

**Review by sleepy-monax — APPROVED** [review]

**Comment by sleepy-monax** [comment]
I seem ok like that, There is still some edge case to handle but this is good enough for now, nice work :)
